api/apikeys.go

```diff
@@ -0,0 +1,69 @@
+package api
+
+import (
+	"net/http"
+
+	"github.com/SkynetLabs/skynet-accounts/jwt"
+	"github.com/julienschmidt/httprouter"
+)
+
+// userAPIKeyPOST creates a new API key for the user.
+func (api *API) userAPIKeyPOST(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	sub, _, _, err := jwt.TokenFromContext(req.Context())
+	if err != nil {
+		api.WriteError(w, err, http.StatusUnauthorized)
+		return
+	}
+	u, err := api.staticDB.UserBySub(req.Context(), sub, true)
+	if err != nil {
+		api.WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	ak, err := api.staticDB.APIKeyCreate(req.Context(), *u)
+	if err != nil {
+		api.WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	api.WriteJSON(w, ak)
+}
+
+// userAPIKeyGET lists all API keys associated with the user.
+func (api *API) userAPIKeyGET(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	sub, _, _, err := jwt.TokenFromContext(req.Context())
+	if err != nil {
+		api.WriteError(w, err, http.StatusUnauthorized)
+		return
+	}
+	u, err := api.staticDB.UserBySub(req.Context(), sub, true)
+	if err != nil {
+		api.WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	aks, err := api.staticDB.APIKeyList(req.Context(), *u)
+	if err != nil {
+		api.WriteError(w, err, http.StatusInternalServerError)
+		return
+	}
+	api.WriteJSON(w, aks)
+}
+
+// userAPIKeyDELETE removes an API key.
+func (api *API) userAPIKeyDELETE(w http.ResponseWriter, req *http.Request, ps httprouter.Params) {
+	sub, _, _, err := jwt.TokenFromContext(req.Context())
```

Comment:
I think we can simplify this a bit. If you adopt the suggestion to rename `validate` to something like `WithAuthentication` we can return `http.StatusUnauthorized` from the middleware. That would avoid checking this manually in all these handlers.

Furthermore, I noticed that we are fetching a user from the database a lot. In a single request's lifetime I think it happens at least twice in all of the happy flows. Perhaps we should make use of `context.WithValue` and attach the user, or decorate the request object with the user object.

Reply:
I think this aligns with my comment above around middleware that basically passes you the User after implicitly fetching it.

Reply:
This is a great suggestion. Do you mind if we do it as a F/U? I'm afraid it's going to touch far too many lines in the service and while it's going to make it smaller and faster, it's going to make this PR far too large.

F/U: #125